File: apps/api/app/routers/bulk.py
"""Bulk creation endpoints for schools and users."""
import secrets
import string
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from app.database import get_db
from app.models.user import User
from app.models.school import School
from app.auth import require_admin, hash_password
from app.services.email import send_user_request_approval_email

router = APIRouter(prefix="/api/bulk", tags=["bulk"])

def _gen_password(length: int = 10) -> str:
    chars = string.ascii_letters + string.digits + "!@#$"
    return ''.join(secrets.choice(chars) for _ in range(length))

@router.post("/users")
async def bulk_create_users(body: dict, db: AsyncSession = Depends(get_db), _: User = Depends(require_admin)):
    """Bulk create users. body.users = list of user dicts."""
    users_data = body.get("users", [])
    if not users_data:
        raise HTTPException(400, "No user data provided")

    SCHOOL_ROLES = {"school_admin", "school_viewer", "teacher"}
    PLATFORM_ROLES = {"platform_admin", "platform_viewer"}
    ALL_ROLES = SCHOOL_ROLES | PLATFORM_ROLES

    results = []
    for row in users_data:
        try:
            role = row.get("role", "school_viewer")
            if role not in ALL_ROLES:
                raise ValueError(f"Invalid role '{role}'. Must be one of: {', '.join(sorted(ALL_ROLES))}")

            school_id = row.get("schoolId")

            if role in SCHOOL_ROLES and not school_id:
                raise ValueError(f"Role '{role}' requires a schoolId")
            if role in PLATFORM_ROLES:
                school_id = None

            # Validate school exists
            if school_id:
                school = await db.get(School, school_id)
                if not school:
                    raise ValueError(f"School '{school_id}' not found")

            existing = await db.scalar(select(User).where(User.email == row["email"]))
            if existing:
                raise ValueError(f"Email '{row['email']}' already in use")

            password = _gen_password()
            user = User(
                email=row["email"],
                password_hash=hash_password(password),
                name=row["name"],
                role=role,
                school_id=school_id,
                mobile_number=row.get("mobileNumber"),
                is_active=True,
                verified=True,
            )
            db.add(user)
            await db.flush()

            # Send welcome email
            try:
                await send_user_request_approval_email(row["email"], row["name"])
            except Exception:
                pass

            results.append({
                "email": row["email"], "name": row["name"], "role": role,
                "generatedPassword": password, "status": "created"
            })
        except Exception as e:
            results.append({"email": row.get("email", "?"), "status": "error", "error": str(e)})

    await db.commit()
    return {
        "results": results,
        "total": len(results),
        "created": sum(1 for r in results if r.get("status") == "created"),
    }

File: apps/api-node/src/routes/bulk.ts
import type { FastifyInstance } from "fastify";
import { z } from "zod";
import { User } from "../models/index.js";
import { hashPassword } from "../lib/auth.js";
import { genPassword } from "../lib/schools.js";
import { PLATFORM_ROLES } from "../lib/permissions.js";
import { requireRole } from "../plugins/auth.js";

const platformRoles = new Set<string>(PLATFORM_ROLES);

export const registerBulkRoutes = async (app: FastifyInstance): Promise<void> => {
  app.post("/api/bulk/users", { preHandler: requireRole(["admin", "platform_admin", "school_admin"]) }, async (request) => {
    const body = z
      .object({
        users: z.array(
          z.object({
            email: z.string().email(),
            password: z.string().min(6).optional(),
            name: z.string(),
            role: z.string().optional(),
            school_id: z.string().nullable().optional(),
            schoolId: z.string().nullable().optional(),
            mobileNumber: z.string().optional(),
            mobile_number: z.string().optional()
          })
        )
      })
      .parse(request.body);

    const results = [];
    for (const row of body.users) {
      try {
        const role = row.role ?? "school_viewer";
        const password = row.password ?? genPassword();
        const schoolId = platformRoles.has(role) ? null : (row.school_id ?? row.schoolId ?? null);
        const user = await User.create({
          email: row.email.toLowerCase(),
          name: row.name,
          role,
          school_id: schoolId,
          mobile_number: row.mobile_number ?? row.mobileNumber,
          password_hash: await hashPassword(password),
          must_change_password: true
        });
        results.push({
          id: user.id,
          name: user.name,
          email: user.email,
          role: user.role,
          schoolId: user.school_id,
          generatedPassword: password,
          status: "created"
        });
      } catch (error) {
        results.push({ email: row.email, name: row.name, status: "error", error: (error as Error).message });
      }
    }

    const created = results.filter((r) => r.status === "created").length;
    return { results, total: results.length, created, created_count: created };
  });
};
